cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -286,6 +339,44 @@ public void validateProperties(String provisionerName, Map<String, String> prope
     provisioner.validateProperties(properties);
   }
 
+  private Runnable createProvisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner) {
+    ProgramRunId programRunId = taskInfo.getProgramRunId();
+    ProvisionerContext context = new DefaultProvisionerContext(programRunId, taskInfo.getProvisionerProperties(),
+                                                               sparkCompat, createSSHContext(taskInfo.getSshKeyInfo()));
+
+    // TODO: (CDAP-13246) pick up timeout from profile instead of hardcoding
+    return new ProvisionTask(taskInfo, provisioner, context, provisionerNotifier, transactional, datasetFramework, 300);
+  }
+
+  private Runnable createDeprovisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner,
+                                         Consumer<ProgramRunId> taskCleanup) {
+    Map<String, String> properties = taskInfo.getProvisionerProperties();
+    ProvisionerContext context = new DefaultProvisionerContext(taskInfo.getProgramRunId(), properties,
+                                                               sparkCompat, createSSHContext(taskInfo.getSshKeyInfo()));
+    return new DeprovisionTask(taskInfo, provisioner, context, provisionerNotifier,
```

Comment:
Instead of having `taskCleanup` passes to the `DeprovisionTask` and requiring it to do the cleanup, I think it is cleanup to have the cleanup controlled by the provisioning service such that the deprovisioning logic does need to be aware of the cleanup at all. You could simply do

```
DeprovisionTask task = new DeprovisionTask(...);
return () -> {
  task.run();
  cleanup.accept(taskInfo.getProgramRunId());
};
```